File: frontend/src/objects/battle/microservices/results/objects/result/BattleResultCommands.ts
import type {
  BattleId,
  BattleMapLabel,
  BattleModeLabel,
  DurationMillis,
  EpochMillis
} from "../../../../objects/core/BattleCoreScalars";
import type { BattleSurvivalOutcome } from "../../../actors/objects/player/BattleSurvivalOutcome";
import type { Rating } from "../../../actors/objects/player/BattlePlayerRating";
import type { Score } from "../../../actors/objects/player/BattlePlayerStats";
import type {
  BattleHighlightLine,
  BattlePlacement,
  BattlePlayersLine,
  BattleResultLabel,
  BattleTimelineHint,
  RatingDelta
} from "./BattleResultPresentationValues";

export interface BattleResultRecordCommand {
  battleId: BattleId;
  handle: string;
  displayName: string;
  finishedAt: EpochMillis;
  finishedAtLabel: string;
  durationMs: DurationMillis;
  score: Score;
  placement: BattlePlacement | null;
  survivalOutcome: BattleSurvivalOutcome;
  ratingBefore: Rating;
  ratingDelta: RatingDelta;
  ratingAfter: Rating;
  resultLabel: BattleResultLabel;
  modeLabel: BattleModeLabel;
  mapLabel: BattleMapLabel;
  highlightLine: BattleHighlightLine;
  playersLine: BattlePlayersLine;
  timelineHint: BattleTimelineHint;
  currentLoadout: string | null;
}

File: backend/src/main/scala/services/battle/microservices/results/objects/result/BattleResultCommands.scala
package services.battle.microservices.results.objects.result

import services.battle.objects.core.{
  BattleId,
  BattleMapLabel,
  BattleModeLabel,
  DurationMillis,
  EpochMillis
}
import services.battle.microservices.actors.objects.player.{BattleSurvivalOutcome, Rating, Score}
import services.identity.objects.{DisplayName, PlayerHandle}

final case class BattleResultRecordCommand(
  battleId: BattleId,
  handle: PlayerHandle,
  displayName: DisplayName,
  finishedAt: EpochMillis,
  finishedAtLabel: BattleResultFinishedAtLabel,
  durationMs: DurationMillis,
  score: Score,
  placement: Option[BattlePlacement],
  survivalOutcome: BattleSurvivalOutcome,
  ratingBefore: Rating,
  ratingDelta: RatingDelta,
  ratingAfter: Rating,
  resultLabel: BattleResultLabel,
  modeLabel: BattleModeLabel,
  mapLabel: BattleMapLabel,
  highlightLine: BattleHighlightLine,
  playersLine: BattlePlayersLine,
  timelineHint: BattleTimelineHint,
  currentLoadout: Option[BattleResultLoadoutLabel]
)

final case class BattleResultListQuery(
  handle: Option[PlayerHandle],
  battleId: Option[BattleId],
  limit: BattleResultListLimit
)

enum BattleResultRecordValidationError {
  case InvalidHandle
  case VisitorNotAllowed
}
